### PR TITLE
Adding pre/post publish scripts

### LINKED ISSUE #742: Chaining `dotnet publish` and `publish-iis`
For Asp.Net projects the output of `dotnet publish` needs to be tweaked a little bit to enable the project to run on IIS. The crucial part is to modify (or create) the web.config file. 

The idea is to create a tool (project tool) which will run after dotnet publish. The tool will take the folder where the project was published and the application name as parameters.

Open questions are: 
- how to chain the tool to run after `dotnet publish`
- where to get the parameter values from - (it might be possible to get the application name by parsing project.json file but the output folder can be provided by the user so it needs to be provided)

/cc @glennc

### PR DESCRIPTION
**Pull request by moozzyk**
Fixes #742

### CODE REVIEW

File: test/Microsoft.DotNet.Tools.Publish.Tests/Microsoft.DotNet.Tools.Publish.Tests.cs [hunk 1]
```diff
@@ -148,6 +145,34 @@ public void CompilationFailedTest()
...
+
+            // run publish
+            var testProject = GetProjectPath(testAppDir);
+            var publishCommand = new PublishCommand(testProject);
+
+            var result = publishCommand.ExecuteWithCapturedOutput();
+
+            Assert.True(
+                Regex.Match(result.StdOut, "\nprepublish_output( '[^%]+'){5}.+\npostpublish_output( '[^%]+'){5}",
+                    RegexOptions.Singleline).Success);
```
Comment (Sridhar-MS) [inline test/Microsoft.DotNet.Tools.Publish.Tests/Microsoft.DotNet.Tools.Publish.Tests.cs]: We are trying to use FluentAssertions style assertions in the tests. You can see an example for asserting stdout [here](https://github.com/dotnet/cli/blob/02d28537f89ddb3908f6f2db90d68da36f36751e/test/E2E/EndToEndTest.cs#L144).

Can you add support for regex matching in the [assertions file](https://github.com/dotnet/cli/blob/02d28537f89ddb3908f6f2db90d68da36f36751e/test/Microsoft.DotNet.Tools.Tests.Utilities/Assertions/CommandResultAssertions.cs)?

Comment (moozzyk) [inline test/Microsoft.DotNet.Tools.Publish.Tests/Microsoft.DotNet.Tools.Publish.Tests.cs]: OK. I will add support regex.

**Comment by moozzyk** [comment]
@blackdwarf does `dotnet-build` publish anything? If it doesn't then these scripts won't impact it in any way since they are invoked when running `dotnet publish` command. If your question is more general and you are also asking about pre/postcompile scripts - ScriptExecutor will run each script with `cmd /s /c` on Windows and `bash -c` on non-Windows. If there are no scripts it's basically no-op.

**Comment by cdmihai** [comment]
Since it is in Publish, it does not affect it directly. 

However, one could add behaviour to the pre - post publish scripts that might cause build to rebuild each time (i.e. scripts that modify input / output files needed for compilation). For example, if a pre-publish script moves the entire output to somewhere else, a subsequent invocation of Build won't find it and trigger a rebuild.

But I guess that's more of a bad build spec rather than a problem with Build.

**Comment by davidfowl** [comment]
:shipit:!

**Comment by moozzyk** [comment]
:up::calendar:
- added fluent assertion for matching regex
- fixed non-windows builds

**Comment by davidfowl** [comment]
:shipit:

**Comment by anurse** [comment]
:shipit: - Not sure what the build failures are but since most of the build succeeded, it looks like probably CI issues...

**Comment by moozzyk** [comment]
Apparently there are some issues with Ubuntu: https://github.com/dotnet/cli/pull/793#issuecomment-170743596. I am not sure about CentOS - can't even connect to the CI.

**Comment by Sridhar-MS** [comment]
FYI. Jenkins is currently down for maintenance.

**Pull request merged**